[숫자 야구 게임] 리나 미션 제출합니다.

1jeongg opened the pull request:
## 자세한 기능 요구 사항 및 클래스별 설명은 `📁./docs/README.md` 를 참고해주세요

## 구현하면서 고민되었던 점
- **validation check할 때 조건 한 번에 체크하기 vs 나눠서 체크하기** 
   - 저는 일단 사용자에게 어디가 잘못되었는지 알려주기 위해서 3번에 나눠서 체크하도록 설정하였는데 regex 나 다른 내장 함수를 이용하면 더 쉽게 체크할 수 있을 것 같네요
   - 📁 `./util/ValidationChecker`
   
- **출력 부분을 Printer 라는 클래스를 만들어서 전담하도록 했는데 너무 과한 것 같기도 하네요...** 
  -  결과 출력하는 코드가 더러운데 더 좋은 방법이 있다면 공유 부탁드립니다🙏
  - 📁 `./util/Printer`
  
<br>

## 자랑하고 싶은 점
- 사용자의 재시작/종료을 위한 숫자를 입력하면 int -> PlayerState로 바꿔서 저장합니다.
   ➡️ 재시작, 종료를 헷갈리지 않게 하고 확장성 생각했을 때 좋을 것 같네요
   ```kotlin
      enum class PlayerState(val state: Int) {
          IN_PROGRESS(1),
          EXIT(2);
      
          companion object {
              // state의 값에 따라 PlayerState를 선택함
              // ex) value에 2가 들어오면 EXIT
              // state의 값이 없으면 default는 IN_PROGRESS
              infix fun from(value: Int): PlayerState =
                      PlayerState.values().firstOrNull { it.state == value } ?: IN_PROGRESS
          }
      }
   ```
- 안드 개발자라면? 못 참는 Backing Properties 사용 
  ```kotlin
    private var _gameNumber = mutableListOf<Int>()
    val gameNumber get() = _gameNumber
  ```
  >  [Backing properties](https://yeongunheo.tistory.com/entry/%EC%BD%94%ED%8B%80%EB%A6%B0-Backing-Field%EC%99%80-Backing-Properties#Backing%--properties) [참고자료2](https://everyday-develop-myself.tistory.com/m/344)
  >
  > Backing properties는 뒷받침하는 프로퍼티라는 뜻이다. 변경 가능한 프로퍼티를 외부에서 변경이 불가능하도록 막고 싶을 때, Backing properties을 사용할 수 있다. 변경 가능한 값을 외부로부터 보호함으로써 유지보수성을 향상시킬 수 있다.
   > Backing properties를 사용하면 클래스 내 같은 상태를 의미하는 2개의 프로퍼티가 공존하게 된다. 이때 Mutable한 프로퍼티는 private 접근 제어자를 붙이고, Mutable하지 않은 프로퍼티는 외부에 공개한다.

책이 아직 배송 중이라 미션을 먼저 하게 되었네요 ㅎㅎ 

앞으로 잘 부탁드립니다 :) 화이팅♥️

1jeongg committed:
docs: add feature requirements

1jeongg committed:
feat: player's getGameNumber and getExitNumber method implemented

1jeongg committed:
test: Player, ValidationChecker's success, fail test

1jeongg committed:
docs: check TodoList for Player

1jeongg committed:
feat: opponent's getAnswerNumber method implemented

1jeongg committed:
test: Opponent's success, fail test

1jeongg committed:
docs: check TodoList for Opponent

1jeongg committed:
feat: calculator's getResult method implemented

플레이어와 상대방의 숫자를 비교하여 스트라이크, 볼, 낫싱 등의 결과를 계산함

1jeongg committed:
test: calculator's test

- 볼만 있는 경우
- 스트라이크만 있는 경우
- 볼, 스트라이크 둘 다 있는 경우
- 낫싱

1jeongg committed:
docs: check TodoList for calculator

1jeongg committed:
feat: implement Printer and BaseBallSystem - entire game logic

1jeongg committed:
test: get PlayerState by value and printer test

1jeongg committed:
refactor: update test code by changes and make code clean

- Player에서 exitNumber PlayerState로 관리하도록 변경하였기에 테스트 코드도 수정
- OpponentTest에서 조건문 하나 더 추가(중복)
- Validation check 메서드명, 변수명 변경
- Opponent에서 새로운 숫자를 생성할 때 초기화

1jeongg committed:
docs: feature requirements readme file updated

02ggang9 commented on kotlin-baseball/src/main/kotlin/baseball/util/ValidationChecker.kt:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    fun checkExitNumber(input: String){
+        if (input != "1" && input != "2") {
+            throw IllegalArgumentException("1 또는 2를 입력해야 합니다.")
+        }
```
코틀린에서 제공하는 required 함수를 사용해 봤는데 좋더라고요! 이 함수 사용에 대해서 다른 분들의 의견이 궁금합니다~

stopmin replied:
오 여기서 required 쓰는 것은 생각치도 못했네요!! 그런식으로 검증하면 좋을 것 같네요!!

1jeongg replied:
저도 처음 알았네요 ㅎㅎ 훨씬 깔끔해져서 좋을 것 같습니다!

02ggang9 commented on kotlin-baseball/src/main/kotlin/baseball/util/ValidationChecker.kt:
```diff
@@ -0,0 +1,40 @@
+        duplicationCheck()
+    }
+
+    fun checkExitNumber(input: String){
+        if (input != "1" && input != "2") {
+            throw IllegalArgumentException("1 또는 2를 입력해야 합니다.")
+        }
+    }
```
코틀린에서 제공하는 required 함수를 사용해 봤는데 좋더라고요! 이 함수 사용에 대해서 다른 분들의 의견이 궁금합니다~

02ggang9 reviewed (approved):
테스트 코드까지 정말 꼼꼼하시네용 bb 입니다~

코틀린이 제공하는 편의 함수가 넘 많아서 계속 새롭네요..!! 계속 보면서 추가로 리뷰 달도록 하겠습니다~

stopmin commented on kotlin-baseball/src/main/kotlin/baseball/BaseBallSystem.kt:
```diff
@@ -0,0 +1,46 @@
+            var isTurnInProgress = true
+            while (isTurnInProgress){
+                // 2. 플레이어: 게임 숫자 입력
+                printer.getGameNumber()
```
md파일에서 조금 고민하신 것 같은데, 오히려 이렇게 프린터 있는게 더 깔끔한 것 같기도 합니다! 

IO관련한 핸들러(입출력을 모두 관리하는 클래스)가 있는 것도꽤 깔끔해 보일 것 같습니다!

stopmin commented on kotlin-baseball/src/main/kotlin/baseball/Opponent.kt:
```diff
@@ -0,0 +1,19 @@
+class Opponent {
+
+    private var _answerNumber = mutableListOf<Int>()
+    val answerNumber get() = _answerNumber
```
오!! 이렇게 커스텀 게터 사용하시는거 신기합니다! 
읽기랑 쓰기를 나누려고 이렇게 하신건가요? 

mutable한 자료형에서는 저렇게 관리하는게 훨씬 더 좋아보이네요! 배워갑니다👍

1jeongg replied:
원래 8번 코드를 `val gameNumber: List<Int> get() = _gameNumber` 이런식으로 써야 하는데 아무 생각 없이 썼네요 ㅎ

stopmin commented on kotlin-baseball/src/main/kotlin/baseball/data/PlayerState.kt:
```diff
@@ -0,0 +1,14 @@
+    IN_PROGRESS(1),
+    EXIT(2);
+
+    companion object {
```
저는 여태 enum안에 그냥 함수로 집어넣었었는데 companion object안에 infix 함수를 써서 집어넣을 생각을 못했네요!! 
그렇게 하면 훨씬 더 코드 안에서 가독성이 좋아져서 사용한 것인가욤?? 
많이 배워가네요 👍

1jeongg replied:
넹 맞습니다! 아무래도 이 프로그램을 만든 사람이 임의로 1, 2을 설정한거다보니 마음에 안들더라고요!

value -> enum class로 변환할 때 아래처럼 썼다가 더 좋은 방법이 없을까.. 하고 사실 구글링하다가 찾은 겁니다 ㅎㅎ
[참고자료](https://www.baeldung.com/kotlin/enum-find-by-value)

```kotlin
val state = if (value == 1) PlayerState.IN_PROGRESS else PlayerState.EXIT
```

근데 생각해보니까 그냥 MAGIC_NUMBER 로 value를 정의하면 굳이 필요 없을 것 같기는 해요!

어 현경님 그러게요? ㅋㅋㅋㅋ ㅠㅠ 중위 함수 선언해두고 일반 메서드처럼 썼네요,, 사실 이번에 infix를 처음 봐서 익숙하지 않았습니다. 수정할 때 `_state = PlayerState from input.toInt()` 이렇게 바꿀게요!
(Player 의 31번째 코드에 있습니다)

stopmin commented on kotlin-baseball/src/main/kotlin/baseball/util/Printer.kt:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    fun getExitNumber(){
+        println("게임을 새로 시작하려면 1, 종료하려면 2를 입력하세요.")
```
저는 출력 멘트를 enum으로 관리했고 꼭 뭔가 이게 정답인 줄 알았는데 선배님 코드 보니 충분히 그렇게 관리하지 않아도 더 깔끔하고 어쩌면 이 코드가 더 깔끔한 것 같다고 느껴지기도 합니닷!!

stopmin reviewed (commented):
안녕하세요!!
선배님 역시 코틀린 고수시네요 👍 
헛헛 보면서 이때까지 제가 코틀린의 극히 일부만 활용하고 있다고 많이 느꼈습니다. 이때까지 구현하면서 걱정했던 부분들에 대한 것들도 선배님 코드 부분 참고하면 해결할 수 있겠다는 생각이 듭니다!! 
앞으로도 잘부탁드립니다🙂
좋은 코드로 구현하시느라 수고 많으셨습니다🔥

shkisme commented on kotlin-baseball/src/main/kotlin/baseball/util/ValidationChecker.kt:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    fun checkExitNumber(input: String){
+        if (input != "1" && input != "2") {
```
질문) 문자열 비교를 `equals()` 메서드 사용 대신 `!=` 로 하셨는데 자바의 `==` 가 코틀린에서는 `===` 이기 때문에 이렇게 사용이 가능한 것 맞나요?

1jeongg replied:
equals() 메서드와 !=는 동일합니다! 실제로 input.equals("1")을 쓰면 `Call can be replaced with binary operator ` 이라는 경고 메시지가  뜹니다

<img width="335" alt="image" src="https://github.com/BDD-CLUB/kotlin-study/assets/84652886/3d4a4894-4cc3-488b-81c2-93de23bfcb09">

shkisme replied:
오 그렇군요!!! 답변 감사합니다. 진짜 코틀린 짱이네요. 짱짱짱

shkisme commented on kotlin-baseball/src/test/kotlin/baseball/OpponentTest.kt:
```diff
@@ -0,0 +1,22 @@
+import org.junit.jupiter.api.Assertions.assertTrue
+import org.junit.jupiter.api.Test
+
+class OpponentTest : NsTest() {
```
`NsTest`를 상속받지 않아도 되는 테스트 클래스들이 존재하는 것 같습니다!

shkisme commented on kotlin-baseball/src/test/kotlin/baseball/OpponentTest.kt:
```diff
@@ -0,0 +1,22 @@
+    fun `상대방 숫자 선택`() {
+        opponent.getAnswerNumber()
+        val result = opponent.answerNumber
+        val isValid = result.size == 3 && result.all { it in 1..9 }
+                && result.distinct().size == result.size
```
우왕 알아갑니다 👍

shkisme commented on kotlin-baseball/src/test/kotlin/baseball/util/ValidationCheckerTest.kt:
```diff
@@ -0,0 +1,74 @@
+
+    private val validationChecker = ValidationChecker()
+
+    // ----------------  게임 숫자 입력 테스트 -------------------
```
`@Nested`를 사용하여 테스트 클래스들을 묶을 수 있을 것 같습니다!

shkisme reviewed (approved):
수고 많으셨습니다~~!! 리뷰하면서 많이 배워갑니다.
꼼꼼하게 리뷰하다보니 리뷰를 많이 남기게 되었네요...ㅎㅎ 확인 부탁드립니다~

1jeongg commented on kotlin-baseball/src/main/kotlin/baseball/util/Printer.kt:
```diff
@@ -0,0 +1,42 @@
+
+import baseball.data.Result
+
+class Printer {
```
앗 그냥 한번에 묶는게 보기 편할 것 같아서 그렇게 했는데 그냥 최상단에 하는게 좋을 것 같네요!

1jeongg commented on kotlin-baseball/src/main/kotlin/baseball/util/Printer.kt:
```diff
@@ -0,0 +1,42 @@
+    fun getExitNumber(){
+        println("게임을 새로 시작하려면 1, 종료하려면 2를 입력하세요.")
+    }
+}
```
왑 감사합니다! 이거 옛날에 이렇게 써야 하는데.. 생각만 하다가 귀찮아서 대충 넘겼는데 이런게 있다니,, ㅠㅠ

좋네요! 안드로이드 스튜디오랑 IntelliJ에 당장 설정하겠습니다 감사합니다👍

1jeongg commented on kotlin-baseball/src/main/kotlin/baseball/util/Calculator.kt:
```diff
@@ -0,0 +1,25 @@
+
+    // 같은 수가 다른 자리에 있으면 볼
+    private fun getBall(player: MutableList<Int>, opponent: MutableList<Int>) =
+            player.filterIndexed { index, num ->
+                num != opponent[index] && opponent.contains(num)
+            }.count()
```
앗 num != opponent[index]는 다른 자리에 있는지를 확인하려 쓴거라 index가 필요할 것 같아요!

shkisme replied:
헐룽 다시 보니 그렇네요 😂 헷갈렸습니다 휴면 에러 발생 ㅋㅋ 🤣

1jeongg commented on kotlin-baseball/src/main/kotlin/baseball/data/PlayerState.kt:
```diff
@@ -0,0 +1,14 @@
+package baseball.data
+
+enum class PlayerState(val state: Int) {
+    IN_PROGRESS(1),
+    EXIT(2);
```
와 아니 저 이거 카테캠할때 맨날 conflict 나서 짜증났는데

이렇게 선언하면 되군요,, 감사합니다 ㅎㅎ

1jeongg committed:
refactor: make code clean referenced by code reviews

- Magic Numebr 사용(3,1,9)
- Printer 내용 최상단 함수로 변경
- Tailing Commas 사용
- PlayerState infix 함수 잘못된 값 입력시 Exception 발생
- data class에 메서드를 생성해 한국어로 변경
- backing properties type 을 mutable하지 않게 변경
- 코틀린 코딩 컨벤션 적용
